File: mvp_app/MyProject/libtsparser/src/main/java/com/tosmart/xiexuming/libtsparser/section/Section.java
package com.tosmart.xiexuming.libtsparser.section;

import androidx.annotation.NonNull;

import java.util.Arrays;

/**
 * @author xiexuming
 */
public class Section {
    private int tableId;
    private int sectionLength;
    private int currentLength;
    private int sectionSyntaxIndicator;
    private int reserved;
    private byte[] data;

    @NonNull
    @Override
    public String toString() {
        return "Section{" +
                "tableId=" + tableId +
                ", sectionLength=" + sectionLength +
                ", currentLength=" + currentLength +
                ", sectionSyntaxIndicator=" + sectionSyntaxIndicator +
                ", reserved=" + reserved +
                ", data=" + Arrays.toString(data) +
                '}';
    }

    public int getSectionSyntaxIndicator() {
        return sectionSyntaxIndicator;
    }

    public void setSectionSyntaxIndicator(int sectionSyntaxIndicator) {
        this.sectionSyntaxIndicator = sectionSyntaxIndicator;
    }

    public int getReserved() {
        return reserved;
    }

    public void setReserved(int reserved) {
        this.reserved = reserved;
    }

    public int getCurrentLength() {
        return currentLength;
    }

    public void setCurrentLength(int currentLength) {
        this.currentLength = currentLength;
    }

    public void setHeadDate(byte[] sectionHead) {
        this.tableId = sectionHead[0];
        this.sectionSyntaxIndicator = (sectionHead[1] >> 8) & 0x01;
        this.reserved = (sectionHead[1] >> 4) & 0x03;
        this.sectionLength = (sectionHead[1] & 0x0F) << 8 | sectionHead[2] & 0xFF;
        this.data = new byte[sectionLength];
    }

    public byte[] getData() {
        return data;
    }

    public int getTableId() {
        return tableId;
    }

    public void setTableId(int tableId) {
        this.tableId = tableId;
    }

    public int getSectionLength() {
        return sectionLength;
    }

    public void setSectionLength(int sectionLength) {
        this.sectionLength = sectionLength;
    }

    public void setData(byte[] data) {
        this.data = data;
    }
}

File: newpaserapp/libtosmartparse/src/main/java/com/example/libtosmartparse/section/Section.kt
package com.example.libtosmartparse.section

import com.example.libtosmartparse.packet.shl
import com.example.libtosmartparse.packet.shr
import java.util.*
import kotlin.experimental.and

/**
 * @author xxm
 * 2022/7/17
 **/
open class Section {
    private var tableId = 0
    private var sectionLength = 0
    private var currentLength = 0
    private var sectionSyntaxIndicator = 0
    private var reserved = 0
    private lateinit var data: ByteArray

    override fun toString(): String {
        return "Section{" +
                "tableId=" + tableId +
                ", sectionLength=" + sectionLength +
                ", currentLength=" + currentLength +
                ", sectionSyntaxIndicator=" + sectionSyntaxIndicator +
                ", reserved=" + reserved +
                ", data=" + Arrays.toString(data) +
                '}'
    }

    fun setHeadDate(sectionHead: ByteArray) {
        tableId = sectionHead[0].toInt()
        sectionSyntaxIndicator = sectionHead[1] shr 8 and 0x01
        reserved = (sectionHead[1] shr 4) and 0x03
        sectionLength = ((sectionHead[1] and 0x0F) shl 8) or (sectionHead[2].toInt() and 0xFF)

        data = ByteArray(sectionLength)
    }

    fun getSectionSyntaxIndicator(): Int {
        return sectionSyntaxIndicator
    }

    fun setSectionSyntaxIndicator(sectionSyntaxIndicator: Int) {
        this.sectionSyntaxIndicator = sectionSyntaxIndicator
    }

    fun getReserved(): Int {
        return reserved
    }

    fun setReserved(reserved: Int) {
        this.reserved = reserved
    }

    fun getCurrentLength(): Int {
        return currentLength
    }

    fun setCurrentLength(currentLength: Int) {
        this.currentLength = currentLength
    }

    fun getData(): ByteArray? {
        return data
    }

    fun getTableId(): Int {
        return tableId
    }

    fun setTableId(tableId: Int) {
        this.tableId = tableId
    }

    fun getSectionLength(): Int {
        return sectionLength
    }

    fun setSectionLength(sectionLength: Int) {
        this.sectionLength = sectionLength
    }

    open fun setData(data: ByteArray?) {
        this.data = data!!
    }
}